src/main/java/com/mongodb/spark/sql/connector/MongoCatalog.java

```diff
@@ -0,0 +1,422 @@
+    initialized = true;
+    catalogName = name;
+    catalogOptions = options;
+    mongoConnectionProvider = new MongoConnectionProvider(options);
+  }
+
+  /**
+   * Called to get this catalog's name.
+   *
+   * <p>This method is only called after {@link #initialize(String, CaseInsensitiveStringMap)} is
+   * called to pass the catalog's name.
+   */
+  @Override
+  public String name() {
+    assertInitialized();
+    return catalogName;
+  }
+
+  /**
+   * List top-level namespaces (databases) from the catalog.
+   *
+   * <p>MongoDB only has top level namespaces (databases) and does not support nested databases.
+   *
+   * @return an array of multi-part namespace (database) names
+   */
+  @Override
+  public String[][] listNamespaces() {
+    assertInitialized();
+    return mongoConnectionProvider.withClient(
+        client ->
+            client.listDatabaseNames().into(new ArrayList<>()).stream()
+                .filter(MongoCatalog::excludeSystemNamespaces)
+                .map(name -> new String[] {name})
+                .toArray(String[][]::new));
+  }
+
+  /**
+   * List namespaces (databases) in a namespace.
+   *
+   * <p>MongoDB only supports top level namespaces (databases) (databases). So will return all
+   * namespaces (databases) if namespace (database) is empty or any empty array if the namespace
+   * (database) exists.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @return an array of multi-part namespace (database) names
+   * @throws NoSuchNamespaceException If the namespace (database) does not exist
+   */
+  @Override
+  public String[][] listNamespaces(final String[] namespace) throws NoSuchNamespaceException {
+    assertInitialized();
+    if (namespace.length == 0) {
+      return listNamespaces();
+    } else if (!namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException(namespace);
+    }
+    return new String[0][];
+  }
+
+  /**
+   * Load metadata properties for a namespace.
+   *
+   * <p>Returns an empty map if the namespace (database) exists.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @return a string map of properties for the given namespace
+   * @throws NoSuchNamespaceException If the namespace (database) does not exist
+   */
+  @Override
+  public Map<String, String> loadNamespaceMetadata(final String[] namespace)
+      throws NoSuchNamespaceException {
+    assertInitialized();
+    if (!namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException(namespace);
+    }
+    return Collections.emptyMap();
+  }
+
+  /**
+   * Test whether a namespace (database) exists.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @return true if the namespace (database) exists, false otherwise
+   */
+  @Override
+  public boolean namespaceExists(final String[] namespace) {
+    assertInitialized();
+    return Arrays.stream(listNamespaces()).anyMatch(s -> Arrays.equals(namespace, s));
```

Comment:
Could we let the server do the filtering for us by using `com.mongodb.client.ListDatabasesIterable#filter`?

Reply:
Done